File: packages/typescript/core/tests/input-resolution.test.ts
import { describe, it, expect } from "vitest";
import { CapabilityInput, InputResolution, InputMeaning, ResolutionMode, ResolutionBehavior } from "../src/models";

describe("v0.24 input resolution", () => {
  it("parses minimal v0.23-shaped input unchanged", () => {
    const inp = CapabilityInput.parse({ name: "q", type: "string" });
    expect(inp.resolution).toBeNull();
    expect(inp.semantic_type).toBeNull();
    expect(inp.catalog_ref).toBeNull();
    expect(inp.entity_reference).toBe(false);
  });

  it("parses closed_values resolution", () => {
    const inp = CapabilityInput.parse({
      name: "forecast_mode",
      type: "string",
      required: false,
      default: "risk_adjusted",
      allowed_values: ["risk_adjusted", "likely", "best_case"],
      semantic_type: "business_category",
      resolution: { mode: "closed_values", on_missing: "use_default", on_ambiguous: "clarify" },
    });
    expect(inp.resolution?.mode).toBe("closed_values");
  });

  it("parses backend_resolved resolution with catalog_ref", () => {
    const inp = CapabilityInput.parse({
      name: "cohort_ref",
      type: "string",
      semantic_type: "cohort_reference",
      entity_reference: true,
      catalog_ref: "gtm.cohort_catalog",
      resolution: { mode: "backend_resolved", resolver_ref: "gtm.cohort_catalog", on_missing: "clarify" },
    });
    expect(inp.resolution?.resolver_ref).toBe("gtm.cohort_catalog");
    expect(inp.catalog_ref).toBe("gtm.cohort_catalog");
    expect(inp.entity_reference).toBe(true);
  });

  it("rejects unknown mode (schema-level)", () => {
    expect(() => InputResolution.parse({ mode: "not_real" })).toThrow();
  });

  it("rejects unknown behavior (schema-level)", () => {
    expect(() => InputResolution.parse({ mode: "clarify", on_missing: "bogus" })).toThrow();
  });

  it("rejects missing mode (required field)", () => {
    expect(() => InputResolution.parse({})).toThrow();
  });

  it("rejects closed_values without allowed_values", () => {
    expect(() => CapabilityInput.parse({
      name: "x",
      type: "string",
      resolution: { mode: "closed_values" },
    })).toThrow();
  });

  it("rejects use_default without default", () => {
    expect(() => CapabilityInput.parse({
      name: "x",
      type: "string",
      resolution: { mode: "clarify", on_missing: "use_default" },
    })).toThrow();
  });

  it("round-trips JSON", () => {
    const original = {
      name: "cohort_ref",
      type: "string",
      required: true,
      default: null,
      description: "",
      semantic_type: "cohort_reference",
      entity_reference: true,
      allowed_values: null,
      catalog_ref: "gtm.cohort_catalog",
      input_meanings: null,
      resolution: { mode: "backend_resolved", resolver_ref: "gtm.cohort_catalog", on_missing: "clarify", on_ambiguous: null, on_unresolved: null },
    };
    const parsed = CapabilityInput.parse(original);
    const serialized = JSON.parse(JSON.stringify(parsed));
    expect(serialized).toEqual(original);
  });
});

File: packages/go/core/input_resolution_test.go
package core

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestInputResolution_MinimalUnchanged(t *testing.T) {
	// Note on `required` defaulting: Python/TS/Java/C# default missing
	// `required` to true via their respective serializer hooks; Go's plain
	// json.Unmarshal sets bool fields to false when absent. That pre-existing
	// asymmetry is not introduced or fixed by v0.24; do not assert on
	// inp.Required here.
	raw := `{"name":"q","type":"string"}`
	var inp CapabilityInput
	if err := json.Unmarshal([]byte(raw), &inp); err != nil {
		t.Fatalf("v0.23-shaped input should parse: %v", err)
	}
	if inp.Resolution != nil {
		t.Errorf("expected nil Resolution")
	}
	if inp.EntityReference {
		t.Errorf("expected entity_reference=false default")
	}
	if inp.CatalogRef != nil {
		t.Errorf("expected nil CatalogRef")
	}
}

func TestInputResolution_BackendResolved(t *testing.T) {
	raw := `{
		"name":"cohort_ref","type":"string","required":true,
		"semantic_type":"cohort_reference","entity_reference":true,"catalog_ref":"gtm.cohort_catalog",
		"resolution":{"mode":"backend_resolved","resolver_ref":"gtm.cohort_catalog","on_missing":"clarify"}
	}`
	var inp CapabilityInput
	if err := json.Unmarshal([]byte(raw), &inp); err != nil {
		t.Fatalf("parse: %v", err)
	}
	if err := ValidateCapabilityInput(&inp); err != nil {
		t.Fatalf("validate: %v", err)
	}
	if inp.Resolution.Mode != ResolutionModeBackendResolved {
		t.Errorf("mode = %q", inp.Resolution.Mode)
	}
	if *inp.CatalogRef != "gtm.cohort_catalog" {
		t.Errorf("catalog_ref = %q", *inp.CatalogRef)
	}
}

func TestInputResolution_UnknownModeRejectedAtDecode(t *testing.T) {
	raw := `{"name":"x","type":"string","resolution":{"mode":"not_real"}}`
	var inp CapabilityInput
	if err := json.Unmarshal([]byte(raw), &inp); err == nil {
		t.Errorf("expected decode error for unknown mode, got nil")
	}
}

func TestInputResolution_UnknownBehaviorRejectedAtDecode(t *testing.T) {
	raw := `{"name":"x","type":"string","resolution":{"mode":"clarify","on_missing":"bogus"}}`
	var inp CapabilityInput
	if err := json.Unmarshal([]byte(raw), &inp); err == nil {
		t.Errorf("expected decode error for unknown behavior, got nil")
	}
}

func TestInputResolution_MissingModeRejectedAtDecode(t *testing.T) {
	raw := `{"name":"x","type":"string","resolution":{}}`
	var inp CapabilityInput
	if err := json.Unmarshal([]byte(raw), &inp); err == nil {
		t.Errorf("expected decode error for missing resolution.mode, got nil")
	}
}

func TestInputResolution_ClosedValuesRequiresAllowedValues(t *testing.T) {
	raw := `{"name":"x","type":"string","resolution":{"mode":"closed_values"}}`
	var inp CapabilityInput
	_ = json.Unmarshal([]byte(raw), &inp)
	err := ValidateCapabilityInput(&inp)
	if err == nil || !strings.Contains(strings.ToLower(err.Error()), "allowed_values") {
		t.Errorf("expected allowed_values error, got %v", err)
	}
}

func TestInputResolution_UseDefaultRequiresDefault(t *testing.T) {
	raw := `{"name":"x","type":"string","resolution":{"mode":"clarify","on_missing":"use_default"}}`
	var inp CapabilityInput
	_ = json.Unmarshal([]byte(raw), &inp)
	err := ValidateCapabilityInput(&inp)
	if err == nil || !strings.Contains(strings.ToLower(err.Error()), "default") {
		t.Errorf("expected default error, got %v", err)
	}
}

func TestInputResolution_RoundTrip(t *testing.T) {
	original := CapabilityInput{
		Name:            "cohort_ref",
		Type:            "string",
		Required:        true,
		SemanticType:    strPtr("cohort_reference"),
		EntityReference: true,
		CatalogRef:      strPtr("gtm.cohort_catalog"),
		Resolution: &InputResolution{
			Mode:        ResolutionModeBackendResolved,
			ResolverRef: strPtr("gtm.cohort_catalog"),
			OnMissing:   behaviorPtr(ResolutionBehaviorClarify),
		},
	}
	b, err := json.Marshal(original)
	if err != nil {
		t.Fatal(err)
	}
	var rt CapabilityInput
	if err := json.Unmarshal(b, &rt); err != nil {
		t.Fatal(err)
	}
	if rt.Resolution.Mode != original.Resolution.Mode {
		t.Errorf("mode lost in round-trip")
	}
	if *rt.CatalogRef != *original.CatalogRef {
		t.Errorf("catalog_ref lost")
	}
}

func strPtr(s string) *string                              { return &s }
func behaviorPtr(b ResolutionBehavior) *ResolutionBehavior { return &b }
